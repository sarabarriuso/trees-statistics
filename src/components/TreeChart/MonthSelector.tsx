import Slider from '@mui/material/Slider';

interface IMonthSelectorProps {
  description: string;
  defaultMonth: number;
  monthSelectedCallback: (value: number) => void;
}
const MonthSelector: React.FC<IMonthSelectorProps> = ({
  description,
  defaultMonth,
  monthSelectedCallback,
}) => {
  const DEFAULT_CLASSNAME = 'month-selector';

  const marks = [
    {
      value: 1,
      label: 'Jan',
    },
    {
      value: 2,
      label: 'Feb',
    },
    {
      value: 3,
      label: 'Mar',
    },
    {
      value: 4,
      label: 'Apr',
    },
    {
      value: 5,
      label: 'May',
    },
    {
      value: 6,
      label: 'Jun',
    },
    {
      value: 7,
      label: 'Jul',
    },
    {
      value: 8,
      label: 'Aug',
    },
    {
      value: 9,
      label: 'Sep',
    },
    {
      value: 10,
      label: 'Oct',
    },
    {
      value: 11,
      label: 'Nov',
    },
    {
      value: 12,
      label: 'Dec',
    },
  ];

  const handleChange = (
    _event: Event | React.SyntheticEvent<Element, Event>,
    value: number | number[],
  ) => {
    monthSelectedCallback(value as number);
  };

  return (
    <div className={DEFAULT_CLASSNAME}>
      <p>{description}</p>
      <Slider
        aria-label='Day'
        defaultValue={defaultMonth}
        track={false}
        valueLabelDisplay='auto'
        step={1}
        marks={marks}
        min={1}
        max={12}
        onChangeCommitted={handleChange}
      />
    </div>
  );
};

export default MonthSelector;
