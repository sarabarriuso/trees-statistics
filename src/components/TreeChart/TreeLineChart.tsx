import { Line } from 'react-chartjs-2';
import {
  Chart as ChartJS,
  CategoryScale,
  LinearScale,
  PointElement,
  LineElement,
  Title,
  Tooltip,
  Legend,
} from 'chart.js';

import { IPlantedTreeData } from '../../interfaces/IPlantedTreeData';

ChartJS.register(
  CategoryScale,
  LinearScale,
  PointElement,
  LineElement,
  Title,
  Tooltip,
  Legend,
);

interface ITreeBarChartProps {
  chartData: Array<IPlantedTreeData>;
}

const TreeLineChart: React.FC<ITreeBarChartProps> = ({ chartData }) => {
  const DEFAULT_CLASSNAME = 'tree-bar-chart';
  const graphData = {
    labels: chartData.map(
      (dataPoint) => dataPoint.date.toISOString().split('T')[0],
    ),
    datasets: [
      {
        label: 'Trees planted',
        data: chartData.map((dataPoint) => dataPoint.count),
        backgroundColor: '#1976d2',
      },
    ],
  };

  return (
    <div className={DEFAULT_CLASSNAME}>
      <Line
        options={{
          responsive: true,
          plugins: {
            legend: {
              position: 'bottom',
            },
            title: {
              display: true,
              text: 'Trees planted per day',
            },
          },
        }}
        data={graphData}
      />
    </div>
  );
};

export default TreeLineChart;
