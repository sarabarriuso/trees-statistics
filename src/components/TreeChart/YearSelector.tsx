import Slider from '@mui/material/Slider';
interface IMonthSelectorProps {
  description: string;
  defaultYear: number;
  minYear?: number;
  yearSelectedCallback: (value: number) => void;
}
const MonthSelector: React.FC<IMonthSelectorProps> = ({
  description,
  defaultYear,
  minYear,
  yearSelectedCallback,
}) => {
  const DEFAULT_CLASSNAME = 'year-selector';

  const handleChange = (
    _event: Event | React.SyntheticEvent<Element, Event>,
    value: number | number[],
  ) => {
    yearSelectedCallback(value as number);
  };

  const marks = [];
  for (let i = 2019; i <= new Date().getFullYear(); i++) {
    marks.push({ value: i, label: i.toString() });
  }

  return (
    <div className={DEFAULT_CLASSNAME}>
      <p>{description}</p>
      <Slider
        aria-label='Day'
        defaultValue={defaultYear}
        track={false}
        valueLabelDisplay='auto'
        step={1}
        marks={marks}
        min={minYear || 2019}
        max={new Date().getFullYear()}
        onChangeCommitted={handleChange}
      />
    </div>
  );
};

export default MonthSelector;
