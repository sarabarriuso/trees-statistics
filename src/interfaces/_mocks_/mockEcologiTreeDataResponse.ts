import { IEcologiTreeDataResponse } from '../responses/IEcologiTreeDataResponse';

export const mockSuccessEcologiTreeDataResponse: IEcologiTreeDataResponse = {
  responseCode: '200',
  responseText: 'Successful',
  data: [[1, 0]],
};
