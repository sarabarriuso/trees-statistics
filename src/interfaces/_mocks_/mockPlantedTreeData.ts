import { IPlantedTreeData } from '../IPlantedTreeData';

export const mockPlantedTreeDataResponse: Array<IPlantedTreeData> = [
  { count: 125, date: new Date(88453136789) },
  { count: 479, date: new Date(88474436748) },
  { count: 741, date: new Date(88412347789) },
  { count: 239, date: new Date(88474633786) },
  { count: 846, date: new Date(88964633235) },
];
