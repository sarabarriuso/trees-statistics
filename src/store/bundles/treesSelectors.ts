import { IPlantedTreeData } from '../../interfaces/IPlantedTreeData';
import IState from '../state';
import { useStateSelector } from '../useSelector';

export const GetPlantedTreesData = (): Array<IPlantedTreeData> | undefined =>
  useStateSelector((state: IState) => state.treesState.plantedTreesData);
export const GetLoadingPlantedTreesData = (): boolean =>
  useStateSelector((state: IState) => state.treesState.loadingPlantedTreesData);
