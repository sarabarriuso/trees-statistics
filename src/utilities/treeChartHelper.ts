import { IPlantedTreeData } from '../interfaces/IPlantedTreeData';

export function filterChartData(
  fromDate: Date,
  toDate: Date,
  data: Array<IPlantedTreeData>,
): Array<IPlantedTreeData> {
  return data.filter(
    (treeData) => treeData.date >= fromDate && treeData.date <= toDate,
  );
}
