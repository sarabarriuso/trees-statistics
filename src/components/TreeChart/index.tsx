import React, { useState } from 'react';
import { GetPlantedTreesData } from '../../store/bundles/treesSelectors';
import { filterChartData } from '../../utilities/treeChartHelper';
import MonthSelector from './MonthSelector';
import TreeLineChart from './TreeLineChart';
import YearSelector from './YearSelector';

const TreeChart: React.FC = () => {
  const DEFAULT_CLASSNAME = 'tree-chart';

  const [fromMonth, setFromMonth] = useState<number>(1);
  const [toMonth, setToMonth] = useState<number>(1);
  const [fromYear, setFromYear] = useState<number>(2019);
  const [toYear, setToYear] = useState<number>(1);

  const treesData = GetPlantedTreesData();

  //const treesData = useMemo(() => GetPlantedTreesData(), []);
  const treeChartData = filterChartData(
    new Date(fromYear, fromMonth - 1),
    new Date(toYear, toMonth - 1),
    treesData || [],
  );

  return (
    <div className={DEFAULT_CLASSNAME}>
      <div className={`${DEFAULT_CLASSNAME}__filters`}>
        <YearSelector
          description={'From year:'}
          defaultYear={fromYear}
          yearSelectedCallback={(value: number) => setFromYear(value)}
        />
        <MonthSelector
          description={'From month:'}
          defaultMonth={fromMonth}
          monthSelectedCallback={(value: number) => setFromMonth(value)}
        />
        <YearSelector
          description={'To year:'}
          defaultYear={toYear}
          minYear={fromYear}
          yearSelectedCallback={(value: number) => setToYear(value)}
        />
        <MonthSelector
          description={'To month:'}
          defaultMonth={toMonth}
          monthSelectedCallback={(value: number) => setToMonth(value)}
        />
      </div>

      {treesData && <TreeLineChart chartData={treeChartData} />}
    </div>
  );
};

export default TreeChart;
