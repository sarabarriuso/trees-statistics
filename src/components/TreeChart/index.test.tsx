import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import * as TreesSelectors from '../../store/bundles/treesSelectors';
import * as TreeChartHelpers from '../../utilities/treeChartHelper';

import { mockPlantedTreeDataResponse } from '../../interfaces/_mocks_/mockPlantedTreeData';
import TreeChart from '.';
import TreeLineChart from './TreeLineChart';

export default describe('TreeChart Index Tests', () => {
  const getPlantedTreesDataMock = jest.spyOn(
    TreesSelectors,
    'GetPlantedTreesData',
  );
  const filterCharDataMock = jest.spyOn(TreeChartHelpers, 'filterChartData');

  afterEach(() => {
    getPlantedTreesDataMock.mockClear();
    filterCharDataMock.mockClear();
  });

  it('Renders successfully when there is not tree data', () => {
    // Arrange
    getPlantedTreesDataMock.mockReturnValue([]);
    filterCharDataMock.mockReturnValue([]);

    // Act
    const tree = shallow(<TreeChart />);

    //Assert
    expect(toJson(tree)).toMatchSnapshot();
    const treeLineChart = tree.find(TreeLineChart);
    expect(treeLineChart.length).toBe(0);
  });

  it('Renders successfully when there is tree data', () => {
    // Arrange
    getPlantedTreesDataMock.mockReturnValue(mockPlantedTreeDataResponse);
    filterCharDataMock.mockReturnValue(mockPlantedTreeDataResponse);
    // Act
    const tree = shallow(<TreeChart />);

    //Assert
    expect(toJson(tree)).toMatchSnapshot();
    const treeLineChart = tree.find(TreeLineChart);
    expect(treeLineChart.length).toBe(1);
  });

  it('Calls filterChartData with the expected values', () => {
    // Arrange
    getPlantedTreesDataMock.mockReturnValue(mockPlantedTreeDataResponse);
    filterCharDataMock.mockReturnValue(mockPlantedTreeDataResponse);

    // Act
    const tree = shallow(<TreeChart />);

    //Assert
    const treeLineChart = tree.find(TreeLineChart);
    expect(filterCharDataMock).toHaveBeenCalledTimes(1);
    expect(filterCharDataMock).toHaveBeenCalledWith(
      new Date(2019, 0),
      new Date(2019, 11),
      mockPlantedTreeDataResponse,
    );
    expect(treeLineChart.length).toBe(1);
  });
});
