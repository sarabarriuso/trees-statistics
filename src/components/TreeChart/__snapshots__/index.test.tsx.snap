// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeChart Index Tests Renders successfully when there is not tree data 1`] = `
<div
  className="tree-chart"
>
  <h2>
    Trees planted per day
  </h2>
  <p>
     
    Please select a time range:
  </p>
  <div
    className="tree-chart__filters"
  >
    <MonthSelector
      defaultYear={2019}
      description="From year:"
      yearSelectedCallback={[Function]}
    />
    <MonthSelector
      defaultMonth={1}
      description="From month:"
      monthSelectedCallback={[Function]}
    />
    <MonthSelector
      defaultYear={2019}
      description="To year:"
      minYear={2019}
      yearSelectedCallback={[Function]}
    />
    <MonthSelector
      defaultMonth={12}
      description="To month:"
      monthSelectedCallback={[Function]}
    />
  </div>
</div>
`;

exports[`TreeChart Index Tests Renders successfully when there is tree data 1`] = `
<div
  className="tree-chart"
>
  <h2>
    Trees planted per day
  </h2>
  <p>
     
    Please select a time range:
  </p>
  <div
    className="tree-chart__filters"
  >
    <MonthSelector
      defaultYear={2019}
      description="From year:"
      yearSelectedCallback={[Function]}
    />
    <MonthSelector
      defaultMonth={1}
      description="From month:"
      monthSelectedCallback={[Function]}
    />
    <MonthSelector
      defaultYear={2019}
      description="To year:"
      minYear={2019}
      yearSelectedCallback={[Function]}
    />
    <MonthSelector
      defaultMonth={12}
      description="To month:"
      monthSelectedCallback={[Function]}
    />
  </div>
  <TreeLineChart
    chartData={
      Array [
        Object {
          "count": 125,
          "date": 1972-10-20T18:18:56.789Z,
        },
        Object {
          "count": 479,
          "date": 1972-10-21T00:13:56.748Z,
        },
        Object {
          "count": 741,
          "date": 1972-10-20T06:59:07.789Z,
        },
        Object {
          "count": 239,
          "date": 1972-10-21T00:17:13.786Z,
        },
        Object {
          "count": 846,
          "date": 1972-10-26T16:23:53.235Z,
        },
      ]
    }
  />
</div>
`;
